File: QytetetJava/src/qytetetjava/Especulador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package qytetetjava;

/**
 *
 * @author vladislav
 */
public class Especulador extends Jugador {
    private int fianza;
    static int FactorEspeculador = 2;
    
    public Especulador(Jugador jugador, int fianza) {
        super(jugador);
        this.fianza = fianza;
    }
    
    @Override
    protected void irACarcel(Casilla casilla) {
        boolean evitarCarcel = this.pagarFianza(fianza);
        
        if (!evitarCarcel) {
            this.setCasillaActual(casilla);
            this.setEncarcelado(true);
        }        
    }
    
    private boolean pagarFianza(int cantidad) {
        boolean puedePagar = false;
        
        if (this.getSaldo() > fianza) {
            puedePagar = true;
            // se supone que hay que pasar la fianza como parametro
            this.modificarSaldo(-cantidad);
        }
        
        return puedePagar;
    }
    
    @Override
    protected void pagarImpuestos(int cantidad) {
        this.modificarSaldo(cantidad/2);
    }
    
    @Override
    protected Especulador convertirme(int fianza) {
        return this;
    }    

    public int getFianza() {
        return fianza;
    }
    
    @Override
    public int getFactorEspeculador() {
        return FactorEspeculador;
    }
    
    
    
}

File: QytetetRuby/lib/especulador.rb
#encoding :UTF-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class Especulador < Jugador
    attr_reader :fianza
    
    def initialize(jugador, fianza)
      super(jugador.nombre, jugador.encarcelado, jugador.saldo, jugador.carta_libertad, jugador.propiedades, jugador.casilla_actual, 2)
      @fianza = fianza
    end
    
    def pagar_impuestos(cantidad)
      modificar_saldo(cantidad/2)
    end
    
    def convertirme(fianza)
      return self
    end
    
    def ir_a_carcel(casilla)
      evitar_carcel = pagar_fianza(@fianza)
      
      if (!evitar_carcel)
        @casilla_actual = casilla
        @encarcelado = true
      end
    end
    
    def pagar_fianza(cantidad)
      puede_pagar = false
      
      if (@saldo > @fianza)
        puede_pagar = true
        modificar_saldo(-@fianza)
      end
      
      return puede_pagar
    end
    
    protected :pagar_impuestos, :convertirme, :ir_a_carcel
    private :pagar_fianza
  end
end
